### ISSUE TITLE
Suggestion : Add optimization functions, ml algorithms under`machine_learning` folder

### ISSUE THREAD

**Issue by Navaneeth-Sharma** [TheAlgorithms/Rust#578]
I would like to suggest adding Optimization functions, machine learning and deep learning algorithms in this repository. This can be added in a separate folder called `machine_learning` folder. Further this folder can contain various algorithms that are used in machine learning and deep learning. 

Initially these Optimization functions can be added under `machine_learning/optmization/`
- [ ] Gradient Descent
- [ ] Stochastic Gradient Descent (SGD)
- [ ] Mini-Batch Gradient Descent
- [ ] Adam (Adaptive Moment Estimation)
- [ ] RMSprop (Root Mean Square Propagation)
- [ ] AdaGrad (Adaptive Gradient Algorithm)

And these in  `machine_learning/`

- [ ] Linear Regression
- [ ] Logistic Regression
- [ ] Decision Trees
- [ ] Random Forest
- [ ] Support Vector Machines (SVM)
- [ ] K-Nearest Neighbors (K-NN)
- [ ] Naive Bayes
- [ ] K-Means Clustering
- [ ] Principal Component Analysis (PCA)

**Comment by siriak** [comment]
I like the idea, let's do that

**Comment by aryan20s** [comment]
Linear regression has been merged (#579), so requesting you to check that part

**Comment by GreatRSingh** [comment]
@Navaneeth-Sharma @siriak Lets also keep loss functions in that folder. #582

**Comment by GreatRSingh** [comment]
@siriak hey, can you something, so that it doesn't get automatically close. I think it should remain open.

**Comment by sahuishan01** [comment]
Hello I am new to open source contribution, I would love to help in implementing some of the optimisation techniques. Kindly let me know if someone is already working on them, or I can start working with few?

**Comment by GreatRSingh** [comment]
You can start working on them

**Comment by haswanth10** [comment]
Pull request has been raised for K-Nearest Neighbors Algorithm (https://github.com/TheAlgorithms/Rust/pull/723). Request you to please check on that

**Comment by jkauerl** [comment]
Hello, I would love to start contributing to this project. Specifically on optimization algorithms like Stochastic Gradient Descent (SGD) or Mini-Batch Gradient Descent. Are they still possible to work on them?

**Comment by siriak** [comment]
Sure, you can start implementing them

**Comment by PratikFandade** [comment]
Hey, I've implemented the logistic regression under this issue, please review the PR for any changes you want - #832

**Comment by anishk85** [comment]
i added knn ml algorithm in pr #923 
but its failing build i checked that it's failing due to other files.

**Comment by Shaurya3101** [comment]
Thats an amazing idea 💡
